src/Libraries/Microsoft.Extensions.AI.Abstractions/AdditionalPropertiesDictionary{TValue}.cs

```diff
@@ -0,0 +1,249 @@
+
+    /// <inheritdoc />
+    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+    /// <inheritdoc />
+    public bool Remove(string key) => _dictionary.Remove(key);
+
+    /// <inheritdoc />
+    bool ICollection<KeyValuePair<string, TValue>>.Remove(KeyValuePair<string, TValue> item) => ((ICollection<KeyValuePair<string, TValue>>)_dictionary).Remove(item);
+
+    /// <summary>Attempts to extract a typed value from the dictionary.</summary>
+    /// <typeparam name="T">Specifies the type of the value to be retrieved.</typeparam>
+    /// <param name="key">The key to locate.</param>
+    /// <param name="value">
+    /// The value retrieved from the dictionary, if found and successfully converted to the requested type;
+    /// otherwise, the default value of <typeparamref name="T"/>.
+    /// </param>
+    /// <returns>
+    /// <see langword="true"/> if a non-<see langword="null"/> value was found for <paramref name="key"/>
+    /// in the dictionary and converted to the requested type; otherwise, <see langword="false"/>.
+    /// </returns>
+    /// <remarks>
+    /// If a non-<see langword="null"/> value is found for the key in the dictionary, but the value is not of the requested type and is
+    /// an <see cref="IConvertible"/> object, the method attempts to convert the object to the requested type.
+    /// </remarks>
+    public bool TryGetValue<T>(string key, [NotNullWhen(true)] out T? value)
+    {
+        if (TryGetValue(key, out TValue? obj))
+        {
+            switch (obj)
+            {
+                case T t:
+                    // The object is already of the requested type. Return it.
+                    value = t;
+                    return true;
+
+                case IConvertible:
+                    // The object is convertible; try to convert it to the requested type. Unfortunately, there's no
+                    // convenient way to do this that avoids exceptions and that doesn't involve a ton of boilerplate,
+                    // so we only try when the source object is at least an IConvertible, which is what ChangeType uses.
+                    try
+                    {
+                        value = (T)Convert.ChangeType(obj, typeof(T), CultureInfo.InvariantCulture);
+                        return true;
+                    }
+                    catch (Exception e) when (e is ArgumentException or FormatException or InvalidCastException or OverflowException)
+                    {
+                        // Ignore known failure modes.
+                    }
+
+                    break;
+            }
+        }
+
+        // Unable to find the value or convert it to the requested type.
+        value = default;
+        return false;
+    }
+
+    /// <inheritdoc />
+    public bool TryGetValue(string key, [NotNullWhen(true)] out TValue value) => _dictionary.TryGetValue(key, out value!);
```

Comment:
This attribute is no longer correct. It should be MaybeNullWhen(false).

Reply:
Just changing the attribute to `[MaybeNullWhen(false)]` results in build errors, e.g.:

```
Nullability of reference types in type of parameter 'value' of 'bool AdditionalPropertiesDictionary<TValue>.TryGetValue(string key, out TValue value)' doesn't match implicitly implemented member 'bool IDictionary<string, TValue>.TryGetValue(string key, out TValue value)' (possibly because of nullability attributes).
```

I'm not super familiar with this but it looks like it's an API difference across .NET versions. Does the following look right to you?

```cs
#if NET
    /// <inheritdoc />
    public bool TryGetValue(string key, [MaybeNullWhen(false)] out TValue value) => _dictionary.TryGetValue(key, out value);
#else
    /// <inheritdoc />
    public bool TryGetValue(string key, [NotNullWhen(true)] out TValue value) => _dictionary.TryGetValue(key, out value!);
#endif
```

Reply:
> Does the following look right to you?

Unfortunately, no. It means downlevel you could have TryGrtValue return true,  have the value be null (because null was put into the dictionary as a valid object? value), but then not get an nrt warning trying to deference the value.

I'm on my phone right now. For now,  can you add an explicit impl of the interface method,  with whatever suppressions you need to make it compile, separate from the public method annotated correctly? The annotations on the explicitly impl won't matter,  and we can sort it out to try to remove those extra impls later.